### PR TITLE
Further changes that will successfully build osqueryd.exe and osqueryi.exe

### LINKED ISSUE #2007: Runs as a Windows service

### PR DESCRIPTION
**Pull request by yying**
This PR contains the set of changes necessary to produce a `osqueryd.exe` and `osqueryi.exe` on Windows. The ported over shell supports `readline` on POSIX and `linenoise-ng` on Windows and includes table autocompletion

Closes #2007

### CODE REVIEW

**Comment by osqueryer** [comment]
Can one of the osquery admins reply with "ok to test" to kick off a build...

**Comment by yying** [comment]
ok to test

**Comment by PoppySeedPlehzr** [comment]
ok to test

File: osquery/devtools/shell.cpp [hunk 1]
```diff
@@ -125,32 +140,84 @@ static sqlite3_int64 timeOfDay(void) {
...
     getrusage(RUSAGE_SELF, &sBegin);
+#endif
+
     iBegin = timeOfDay();
   }
 }
 
 // Return the difference of two time_structs in seconds
+#ifdef WIN32
+static double timeDiff(ULARGE_INTEGER *pStart, ULARGE_INTEGER *pEnd) {
```
Comment (PoppySeedPlehzr) [inline osquery/devtools/shell.cpp]: Should we consider making an issue to build out abstraction classes for small utilitarian type functions such as these? It seems that small things like this keep showing up, and I wonder if it'd be nice for us to have all of these small helper functions centralized and abstracted?

**Comment by osqueryer** [comment]
Build origin/pr/2283/merge (Job: 2925) is good! :+1:

File: osquery/devtools/shell.cpp [hunk 2]
```diff
@@ -125,32 +140,84 @@ static sqlite3_int64 timeOfDay(void) {
...
 // End the timer and print the results.
 static void endTimer(void) {
   if (enableTimer) {
-    struct rusage sEnd;
     sqlite3_int64 iEnd = timeOfDay();
+#ifdef WIN32
+    ULARGE_INTEGER sKernelTimeEnd = {0};
+    ULARGE_INTEGER sUserTimeEnd = {0};
+
+    FILETIME ftCreation = {0};
```
Comment (theopolis) [inline osquery/devtools/shell.cpp]: And there's a bit of duplication here too.

File: osquery/tables/system/cpuid.cpp [hunk 3]
```diff
@@ -45,10 +45,14 @@ std::map<int, std::vector<FeatureDef_t> > kCPUFeatures = {
      }},
 };
 
-static inline void cpuid(unsigned int eax, unsigned int ecx, int regs[4]) {
+static inline void cpuid(size_t eax, size_t ecx, int regs[4]) {
+#if defined(WIN32)
+  __cpuidex((int*)regs, (int)eax, (int)ecx);
```
Comment (theopolis) [inline osquery/tables/system/cpuid.cpp]: Are you spying on my branches? ;)

File: osquery/tables/system/uptime.cpp [hunk 4]
```diff
@@ -8,45 +8,49 @@
...
+  if (sysinfo(&sys_info) != 0) {
+    return -1;
+  }
 
-    return sys_info.uptime;
-  #endif
+  return sys_info.uptime;
+#elif defined(WIN32)
+  auto boot_time = std::chrono::milliseconds(GetTickCount64());
+  return std::chrono::duration_cast<std::chrono::seconds>(boot_time).count();
```
Comment (theopolis) [inline osquery/tables/system/uptime.cpp]: More creeping detected: https://github.com/theopolis/osquery/commit/e2af071f4dd1f272769ac41d2cd9476cf5591ed8#diff-2be235f7d43e3f0f5e56d12b4d134d7eR52

File: osquery/tables/system/cpuid.cpp (hunk 3 above)
Comment (yying) [inline osquery/tables/system/cpuid.cpp]: Is `cpuid.cpp` ready yet? `osquery_distributed` appears to require it, which is why I included a copy from your `win10-build-test` branch.

Comment (theopolis) [inline osquery/tables/system/cpuid.cpp]: I don't think there's a requirement? Where do you see that?

**Comment by theopolis** [comment]
Do you mind backing off the e50a976d07d24f5cbc3d98d00f48844d3e3a0b29 commit? Then I can merge my changes for the two table ports separately.

**Comment by yying** [comment]
Yea, sure.

**Comment by osqueryer** [comment]
Build origin/pr/2283/merge (Job: 2928) is good! :+1:

**Pull request merged**